src/particles/diagnostics/DiagnosticOutput.cpp

```diff
@@ -0,0 +1,63 @@
+/* Copyright 2021 Axel Huebl, Chad Mitchell
+ *
+ * This file is part of ImpactX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#include "DiagnosticOutput.H"
+
+#include <AMReX_Extension.H>  // for AMREX_RESTRICT
+#include <AMReX_REAL.H>       // for ParticleReal
+#include <AMReX_Print.H>      // for PrintToFile
+
+namespace impactx
+{
+namespace diagnostics {
+    void DiagnosticOutput(ImpactXParticleContainer &pc,
+                                  OutputType const &otype) {
+        using namespace amrex::literals; // for _rt and _prt
+
+        // loop over refinement levels
+        int const nLevel = pc.maxLevel();
+        for (int lev = 0; lev < nLevel; ++lev) {
+            // loop over all particle boxes
+            using ParIt = ImpactXParticleContainer::iterator;
+            for (ParIt pti(pc, lev); pti.isValid(); ++pti) {
+                const int np = pti.numParticles();
+
+                // preparing access to particle data: AoS
+                using PType = ImpactXParticleContainer::ParticleType;
+                auto &aos = pti.GetArrayOfStructs();
+                PType *AMREX_RESTRICT aos_ptr = aos().dataPtr();
+
+                // preparing access to particle data: SoA of Reals
+                auto &soa_real = pti.GetStructOfArrays().GetRealData();
+                amrex::ParticleReal *const AMREX_RESTRICT part_px = soa_real[RealSoA::ux].dataPtr();
+                amrex::ParticleReal *const AMREX_RESTRICT part_py = soa_real[RealSoA::uy].dataPtr();
+                amrex::ParticleReal *const AMREX_RESTRICT part_pt = soa_real[RealSoA::pt].dataPtr();
+
+                if( otype == OutputType::PrintParticles) {
+                    amrex::ParallelFor(np, [=] AMREX_GPU_DEVICE(long i) {
+                        // access AoS data such as positions and cpu/id
+                        PType &p = aos_ptr[i];
+                        amrex::ParticleReal const x = p.pos(0);
+                        amrex::ParticleReal const y = p.pos(1);
+                        amrex::ParticleReal const t = p.pos(2);
+
+                        // access SoA Real data
+                        amrex::ParticleReal &px = part_px[i];
+                        amrex::ParticleReal &py = part_py[i];
+                        amrex::ParticleReal &pt = part_pt[i];
+
+                        // write particle data to file
+                        amrex::PrintToFile("output_beam.txt") << x << " " << y << " ";
```

Comment:
Managed memory: copy from GPU to CPU
https://github.com/ECP-WarpX/WarpX/blob/fe971e82a49a1c5101e5ef48cce849460dcb598d/Source/Diagnostics/ReducedDiags/FieldProbe.cpp#L515-L531

The latter array can then be printed to file after that loop

Reply:
Later we do explicit copies to avoid managed memory: https://github.com/AMReX-Codes/amrex/blob/6e27136f76280e565d30ff654031d1b1ac683171/Src/Particle/AMReX_ParticleIO.H#L608

Reply:
if we want to do this properly, there is an easy way that I forgot about!
```C++
// create a host-side particle buffer
auto tmp = amrex::ParticleBuffer::getTmpPC<amrex::PinnedArenaAllocator>(pc);

// copy device-to-host
using SrcData = WarpXParticleContainer::ParticleTileType::ConstParticleTileDataType;
bool const local = true;
tmp.copyParticles(*pc, local);
```

The object in `tmp` behave just like our particle container `pc`.